### ISSUE TITLE
Missing required position argument in net.py line 434.

### ISSUE THREAD

**Issue by ezeike** [rethinkdb/rethinkdb#4402]
I am coming across an error in my application that results in the following exception:
```
File "/home/ubuntu/vclientapp/lib/python3.4/site-packages/rethinkdb/net.py", line 434, in _read_response
    self.close(noreply_wait=False)
TypeError: close() missing 1 required positional argument: 'token'
```
Looking at that class I can see self.close requires a token positional argument. I changed line 434 to
```
self.close(noreply_wait=False, token=res_token)
```

and that particular error went away. 

I cannot figure out what is causing the "Unexpected response received". It only happens when my application is loaded in it's entirety (it's a django app). 

If I send the problematic HTTP requests separately (using firebug, etc) it works fine.

The error does not happen on the same query every time either. They are all simple queries also.

The rethinkdb server prints no error messages, I can't figure out how to track this down.

I am using rethinkdb 2.0.3 and the python drivers 2.0.0-post2 with python3.

With pip there's also a conflict of some sort with versions:
```
rethinkdb                     - This package provides the Python driver library for the RethinkDB database server.
  INSTALLED: 2.0.0.post2
  LATEST:    2.0.0-2

```

**Comment by danielmewes** [comment]
Thanks for the issue report and for investigating the problem @ezeike !

So it seems like there are two problems here.
Obviously we should fix the problem with `close`.
I think what's happening here is actually that it's calling the wrong `close` method, namely one that's meant to be used for closing cursors, while it looks like that it means to call the `close` method to close the connection.

@Tryneus could you take a look at that please?

The other question is why you got an "Unexpected response received" error in the first place.

@AtnNn Do you know where the "2.0.0.post2" pip driver version is coming from? Is this actually a version that we've used at some point?

**Comment by danielmewes** [comment]
Well after looking at the code it doesn't seem to be calling the cursor `close` as I had speculated, but it's not calling `Connection.close` either. Instead it seems to be calling `ConnectionInstance.close` which needs a token (though that token should be a new token I think, not the token of the read response).

**Comment by AtnNn** [comment]
I don't know where the `2.0.0.post2` version comes from.

**Comment by danielmewes** [comment]
@ezeike did you install the RethinkDB driver manually at any point?
I recommend updating to the `2.0.0-2` version (`pip install --upgrade rethinkdb` is probably enough), since we are not sure which version the `2.0.0.post2` one actually is.

**Comment by ezeike** [comment]
@danielmewes I don't believe I've ever installed the rethinkdb driver manually.

I have no system (ubuntu) rethink packages installed. After running pip uninstall I did the following:

```
find / | grep rethink
grep -r rethink /home/ubuntu (my dev directory)
```
To ensure no files are hanging around.

The post2 thing happens when I install older versions as well.
```
(vclientapp) ubuntu@clientappdev:~$ pip install rethinkdb
Collecting rethinkdb
  Using cached rethinkdb-2.0.0-2.tar.gz
Installing collected packages: rethinkdb
  Running setup.py install for rethinkdb
Successfully installed rethinkdb-2.0.0.post2

(vclientapp) ubuntu@clientappdev:~$ pip install rethinkdb==2.0.0-1
Collecting rethinkdb==2.0.0-1
  Using cached rethinkdb-2.0.0-1.tar.gz
Installing collected packages: rethinkdb
  Running setup.py install for rethinkdb
Successfully installed rethinkdb-2.0.0.post1
```
I also tried 1.16.0-4 and it resulted in post4.

**Comment by danielmewes** [comment]
Interesting. That seems to be a `pip` thing then...

**Comment by ezeike** [comment]
Yeah it does. Looks like I'm getting the correct version of the driver though.

Anyone know what I can do about the missing token parameter? I'd like to get that fixed so I can keep investigating my other problems. I looked at the code but I couldn't figure out how to generate a new token in that spot.

**Comment by Tryneus** [comment]
Ok, looking into this now.

**Comment by Tryneus** [comment]
For the record, the line can safely be changed to `self.close(noreply_wait=False, token=None)`, because the token is only used if we send a noreply_wait.  Looking into why we might get an unexpected response.

I suspect this might be because of a cursor being closed while a response is on the wire, but we're releasing it from the cursor cache too early.

**Comment by Tryneus** [comment]
@ezeike, are you by chance using the connection (or cursors from the same connection) from multiple threads?  The python driver classes are not thread-safe and this can cause unexpected behavior.

**Comment by Tryneus** [comment]
The fix for the incorrect parameters has been approved and merged to `next` in commit 2576752, review was 3005.  Will be in release 2.1.  At this point, since I haven't been able to reproduce any `unexpected response` errors, and I haven't heard back from @ezeike, I'm going to assume that error was due to using a single connection concurrently across threads.  Please reopen if this is not the case.
